File: splp-php/src/Core/Logging/CassandraLogger.php
<?php

declare(strict_types=1);

namespace Splp\Messaging\Core\Logging;

use Splp\Messaging\Exceptions\MessagingException;
use Cassandra\Cluster;
use Cassandra\Session;
use Cassandra\ExecutionOptions;

/**
 * Cassandra Logger
 * 
 * Provides logging and tracing functionality with Cassandra backend
 * Compatible with the TypeScript/Bun implementation
 */
class CassandraLogger
{
    private Cluster $cluster;
    private Session $session;
    private array $config;
    private bool $initialized = false;

    public function __construct(array $config)
    {
        $this->config = $config;
        $this->initializeCluster();
    }

    /**
     * Initialize Cassandra connection and create tables
     */
    public function initialize(): void
    {
        try {
            $this->session = $this->cluster->connect();
            $this->createKeyspace();
            $this->createTables();
            $this->initialized = true;
        } catch (\Exception $e) {
            throw new MessagingException('Failed to initialize Cassandra logger: ' . $e->getMessage());
        }
    }

    /**
     * Log entry to Cassandra
     */
    public function log(array $entry): void
    {
        if (!$this->initialized) {
            throw new MessagingException('Cassandra logger not initialized');
        }

        try {
            $statement = $this->session->prepare(
                "INSERT INTO {$this->config['keyspace']}.message_logs 
                 (request_id, timestamp, type, topic, payload, success, error, duration_ms) 
                 VALUES (?, ?, ?, ?, ?, ?, ?, ?)"
            );

            $this->session->execute($statement, [
                'arguments' => [
                    $entry['request_id'],
                    $entry['timestamp'],
                    $entry['type'],
                    $entry['topic'],
                    $entry['payload'] ? json_encode($entry['payload']) : null,
                    $entry['success'] ?? null,
                    $entry['error'] ?? null,
                    $entry['duration_ms'] ?? null
                ]
            ]);
        } catch (\Exception $e) {
            error_log('Failed to log to Cassandra: ' . $e->getMessage());
        }
    }

    /**
     * Get logs by request ID
     */
    public function getByRequestId(string $requestId): array
    {
        $statement = $this->session->prepare(
            "SELECT * FROM {$this->config['keyspace']}.message_logs 
             WHERE request_id = ? ORDER BY timestamp DESC"
        );

        $result = $this->session->execute($statement, [
            'arguments' => [$requestId]
        ]);

        return $this->formatResults($result);
    }

    /**
     * Get logs by topic
     */
    public function getByTopic(string $topic, int $limit = 100): array
    {
        $statement = $this->session->prepare(
            "SELECT * FROM {$this->config['keyspace']}.message_logs 
             WHERE topic = ? ORDER BY timestamp DESC LIMIT ?"
        );

        $result = $this->session->execute($statement, [
            'arguments' => [$topic, $limit]
        ]);

        return $this->formatResults($result);
    }

    /**
     * Get logs by time range
     */
    public function getByTimeRange(\DateTime $startTime, \DateTime $endTime): array
    {
        $statement = $this->session->prepare(
            "SELECT * FROM {$this->config['keyspace']}.message_logs 
             WHERE timestamp >= ? AND timestamp <= ? ORDER BY timestamp DESC"
        );

        $result = $this->session->execute($statement, [
            'arguments' => [$startTime, $endTime]
        ]);

        return $this->formatResults($result);
    }

    /**
     * Health check
     */
    public function healthCheck(): array
    {
        $startTime = microtime(true);
        
        try {
            $statement = $this->session->prepare("SELECT now() FROM system.local");
            $this->session->execute($statement);
            
            return [
                'status' => 'healthy',
                'responseTime' => (microtime(true) - $startTime) * 1000,
                'lastChecked' => new \DateTime(),
                'metadata' => [
                    'keyspace' => $this->config['keyspace']
                ]
            ];
        } catch (\Exception $e) {
            return [
                'status' => 'unhealthy',
                'message' => $e->getMessage(),
                'responseTime' => (microtime(true) - $startTime) * 1000,
                'lastChecked' => new \DateTime()
            ];
        }
    }

    /**
     * Close connection
     */
    public function close(): void
    {
        if ($this->session) {
            $this->session->close();
        }
        
        if ($this->cluster) {
            $this->cluster->close();
        }
    }

    /**
     * Initialize Cassandra cluster
     */
    private function initializeCluster(): void
    {
        $this->cluster = \Cassandra::cluster()
            ->withContactPoints($this->config['contactPoints'])
            ->withDefaultConsistency(\Cassandra::CONSISTENCY_ONE)
            ->build();
    }

    /**
     * Create keyspace if not exists
     */
    private function createKeyspace(): void
    {
        $keyspace = $this->config['keyspace'];
        
        $this->session->execute(
            "CREATE KEYSPACE IF NOT EXISTS {$keyspace} 
             WITH REPLICATION = {
                 'class': 'SimpleStrategy',
                 'replication_factor': 1
             } AND DURABLE_WRITES = true"
        );
    }

    /**
     * Create tables if not exist
     */
    private function createTables(): void
    {
        $keyspace = $this->config['keyspace'];
        
        // Message logs table with 1 week TTL
        $this->session->execute(
            "CREATE TABLE IF NOT EXISTS {$keyspace}.message_logs (
                request_id text,
                timestamp timestamp,
                type text,
                topic text,
                payload text,
                success boolean,
                error text,
                duration_ms int,
                PRIMARY KEY (request_id, timestamp)
            ) WITH CLUSTERING ORDER BY (timestamp DESC)
            AND default_time_to_live = 604800"
        );

        // Create index for topic queries
        $this->session->execute(
            "CREATE INDEX IF NOT EXISTS ON {$keyspace}.message_logs (topic)"
        );
    }

    /**
     * Format Cassandra results
     */
    private function formatResults($result): array
    {
        $logs = [];
        
        foreach ($result as $row) {
            $logs[] = [
                'request_id' => $row['request_id'],
                'timestamp' => $row['timestamp'],
                'type' => $row['type'],
                'topic' => $row['topic'],
                'payload' => $row['payload'] ? json_decode($row['payload'], true) : null,
                'success' => $row['success'],
                'error' => $row['error'],
                'duration_ms' => $row['duration_ms']
            ];
        }
        
        return $logs;
    }
}

File: splp-net/Logging/CassandraLogger.cs
using Cassandra;
using Microsoft.Extensions.Logging;
using SplpNet.Models;
using System.Text.Json;

namespace SplpNet.Logging;

/// <summary>
/// Cassandra logger for distributed tracing and request/response logging
/// </summary>
public class CassandraLogger : IDisposable
{
    private readonly CassandraConfig _config;
    private readonly ILogger<CassandraLogger>? _logger;
    private ISession? _session;
    private ICluster? _cluster;
    private bool _disposed;

    // Prepared statements for better performance
    private PreparedStatement? _insertLogStatement;
    private PreparedStatement? _selectByRequestIdStatement;
    private PreparedStatement? _selectByTimeRangeStatement;

    public CassandraLogger(CassandraConfig config, ILogger<CassandraLogger>? logger = null)
    {
        _config = config ?? throw new ArgumentNullException(nameof(config));
        _logger = logger;
    }

    /// <summary>
    /// Initializes the Cassandra connection and creates necessary tables
    /// </summary>
    public async Task InitializeAsync()
    {
        if (_session != null)
            return;

        try
        {
            _cluster = Cluster.Builder()
                .AddContactPoints(_config.ContactPoints)
                .Build();

            _session = await _cluster.ConnectAsync();
            _logger?.LogInformation("Connected to Cassandra cluster");

            // Create keyspace if it doesn't exist
            await CreateKeyspaceAsync();

            // Use the keyspace
            await _session.ExecuteAsync(new SimpleStatement($"USE {_config.Keyspace}"));

            // Create tables if they don't exist
            await CreateTablesAsync();

            // Prepare statements
            await PrepareStatementsAsync();

            _logger?.LogInformation("Cassandra logger initialized successfully");
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to initialize Cassandra logger");
            throw;
        }
    }

    /// <summary>
    /// Logs a request or response entry
    /// </summary>
    public async Task LogAsync(LogEntry entry)
    {
        if (_session == null || _insertLogStatement == null)
            throw new InvalidOperationException("CassandraLogger not initialized. Call InitializeAsync first.");

        try
        {
            var payloadJson = JsonSerializer.Serialize(entry.Payload);
            
            var boundStatement = _insertLogStatement.Bind(
                entry.RequestId,
                entry.Timestamp,
                entry.Type,
                entry.Topic,
                payloadJson,
                entry.Success,
                entry.Error,
                entry.DurationMs
            );

            await _session.ExecuteAsync(boundStatement);
            
            _logger?.LogDebug("Logged entry for request {RequestId}", entry.RequestId);
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to log entry for request {RequestId}", entry.RequestId);
            throw;
        }
    }

    /// <summary>
    /// Gets all log entries for a specific request ID
    /// </summary>
    public async Task<List<LogEntry>> GetLogsByRequestIdAsync(string requestId)
    {
        if (_session == null || _selectByRequestIdStatement == null)
            throw new InvalidOperationException("CassandraLogger not initialized. Call InitializeAsync first.");

        try
        {
            var boundStatement = _selectByRequestIdStatement.Bind(requestId);
            var rowSet = await _session.ExecuteAsync(boundStatement);

            var entries = new List<LogEntry>();
            foreach (var row in rowSet)
            {
                entries.Add(MapRowToLogEntry(row));
            }

            return entries.OrderBy(e => e.Timestamp).ToList();
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to get logs for request {RequestId}", requestId);
            throw;
        }
    }

    /// <summary>
    /// Gets log entries within a time range
    /// </summary>
    public async Task<List<LogEntry>> GetLogsByTimeRangeAsync(DateTime startTime, DateTime endTime)
    {
        if (_session == null || _selectByTimeRangeStatement == null)
            throw new InvalidOperationException("CassandraLogger not initialized. Call InitializeAsync first.");

        try
        {
            var boundStatement = _selectByTimeRangeStatement.Bind(startTime, endTime);
            var rowSet = await _session.ExecuteAsync(boundStatement);

            var entries = new List<LogEntry>();
            foreach (var row in rowSet)
            {
                entries.Add(MapRowToLogEntry(row));
            }

            return entries.OrderBy(e => e.Timestamp).ToList();
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to get logs for time range {StartTime} - {EndTime}", startTime, endTime);
            throw;
        }
    }

    private async Task CreateKeyspaceAsync()
    {
        var createKeyspaceCql = $@"
            CREATE KEYSPACE IF NOT EXISTS {_config.Keyspace}
            WITH REPLICATION = {{
                'class': 'SimpleStrategy',
                'replication_factor': 1
            }}";

        await _session!.ExecuteAsync(new SimpleStatement(createKeyspaceCql));
        _logger?.LogInformation("Keyspace {Keyspace} created or already exists", _config.Keyspace);
    }

    private async Task CreateTablesAsync()
    {
        // Create logs table with TTL (1 week = 604800 seconds)
        var createLogTableCql = @"
            CREATE TABLE IF NOT EXISTS message_logs (
                request_id text,
                timestamp timestamp,
                type text,
                topic text,
                payload text,
                success boolean,
                error text,
                duration_ms bigint,
                PRIMARY KEY (request_id, timestamp)
            ) WITH CLUSTERING ORDER BY (timestamp ASC)
            AND default_time_to_live = 604800";

        await _session!.ExecuteAsync(new SimpleStatement(createLogTableCql));

        // Create index on timestamp for time range queries
        var createTimestampIndexCql = @"
            CREATE INDEX IF NOT EXISTS idx_message_logs_timestamp 
            ON message_logs (timestamp)";

        await _session.ExecuteAsync(new SimpleStatement(createTimestampIndexCql));

        _logger?.LogInformation("Cassandra tables created or already exist");
    }

    private async Task PrepareStatementsAsync()
    {
        // Insert statement
        var insertCql = @"
            INSERT INTO message_logs (request_id, timestamp, type, topic, payload, success, error, duration_ms)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)";
        _insertLogStatement = await _session!.PrepareAsync(insertCql);

        // Select by request ID
        var selectByRequestIdCql = @"
            SELECT * FROM message_logs WHERE request_id = ?";
        _selectByRequestIdStatement = await _session.PrepareAsync(selectByRequestIdCql);

        // Select by time range
        var selectByTimeRangeCql = @"
            SELECT * FROM message_logs WHERE timestamp >= ? AND timestamp <= ? ALLOW FILTERING";
        _selectByTimeRangeStatement = await _session.PrepareAsync(selectByTimeRangeCql);

        _logger?.LogDebug("Prepared statements created");
    }

    private static LogEntry MapRowToLogEntry(Row row)
    {
        var payloadJson = row.GetValue<string>("payload");
        var payload = string.IsNullOrEmpty(payloadJson) 
            ? new object() 
            : JsonSerializer.Deserialize<object>(payloadJson) ?? new object();

        return new LogEntry
        {
            RequestId = row.GetValue<string>("request_id"),
            Timestamp = row.GetValue<DateTimeOffset>("timestamp").DateTime,
            Type = row.GetValue<string>("type"),
            Topic = row.GetValue<string>("topic"),
            Payload = payload,
            Success = row.GetValue<bool?>("success"),
            Error = row.GetValue<string?>("error"),
            DurationMs = row.GetValue<long?>("duration_ms")
        };
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _session?.Dispose();
            _cluster?.Dispose();
            _disposed = true;
        }
        GC.SuppressFinalize(this);
    }
}
